controller/konnect/ops/ops_controlplane.go

```diff
@@ -113,7 +131,104 @@ func updateControlPlane(
 	}
 
 	cp.Status.SetKonnectID(*resp.ControlPlane.ID)
+
+	if err := setGroupMembers(ctx, cl, cp, sdkGroups); err != nil {
+		SetKonnectEntityProgrammedConditionFalse(cp, conditions.KonnectGatewayControlPlaneProgrammedReasonFailedToSetControlPlaneGroupMembers, err.Error())
+		return err
+	}
+
 	SetKonnectEntityProgrammedCondition(cp)
 
 	return nil
 }
+
+func setGroupMembers(
+	ctx context.Context,
+	cl client.Client,
+	cp *konnectv1alpha1.KonnectGatewayControlPlane,
+	sdkGroups ControlPlaneGroupSDK,
+) error {
+	if len(cp.Spec.Members) == 0 ||
+		cp.Spec.ClusterType == nil ||
+		*cp.Spec.ClusterType != sdkkonnectcomp.ClusterTypeClusterTypeControlPlaneGroup {
+		return nil
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(cp.Spec.Members))
+	ch := make(chan string)
+	chErr := make(chan error)
+	errs := make([]error, 0)
+	for _, member := range cp.Spec.Members {
+		go func() {
+			var memberCP konnectv1alpha1.KonnectGatewayControlPlane
+			err := cl.Get(ctx, client.ObjectKey{Namespace: cp.Namespace, Name: member.Name}, &memberCP)
+			if err != nil {
+				if k8serrors.IsNotFound(err) {
+					chErr <- err
+					return
+				}
+				chErr <- fmt.Errorf("failed to get control plane group member %s: %w", member.Name, err)
+				return
+			}
+
+			if memberCP.GetKonnectID() == "" {
+				chErr <- fmt.Errorf("control plane group %s member %s has no Konnect ID", cp.Name, member.Name)
+				return
+			}
+			ch <- memberCP.GetKonnectID()
+		}()
+	}
+
+	members := make([]sdkkonnectcomp.Members, 0, len(cp.Spec.Members))
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case err, ok := <-chErr:
+				if !ok {
+					return
+				}
+				errs = append(errs, err)
+				wg.Done()
```

Comment:
Wouldn't it make more sense to call `wg.Done()` as a deferred function in the producer?

Reply:
I thought about that but then I realized that there's a chance that we can call `.Done()` `n` times thus making the `.Wait()` unblock while the goroutine receiving on the channel could still write to the `members` slice while line 208 would read from the slice.

I'm pretty sure there is very little chance for this to happen but I believe it's still possible. WDYT?

Reply:
Maybe we could simplify this a little bit by not running the consumer in a goroutine, but in the main thread like so:
```go
	// Spawn producer goroutines that will fetch the Konnect ID of each member. They will push the IDs to the result
	// channel or any errors encountered to the error channel.
	var (
		ch    = make(chan string)
		chErr = make(chan error)
		wg    = sync.WaitGroup{}
	)
	wg.Add(len(cp.Spec.Members))
	for _, member := range cp.Spec.Members {
		go func() {
			defer wg.Done()

			var memberCP konnectv1alpha1.KonnectGatewayControlPlane
			err := cl.Get(ctx, client.ObjectKey{Namespace: cp.Namespace, Name: member.Name}, &memberCP)
			if err != nil {
				chErr <- fmt.Errorf("failed to get control plane group member %s: %w", member.Name, err)
				return
			}

			if memberCP.GetKonnectID() == "" {
				chErr <- fmt.Errorf("control plane group %s member %s has no Konnect ID", cp.Name, member.Name)
				return
			}
			ch <- memberCP.GetKonnectID()
		}()
	}
	// Wait for all producer goroutines to finish and close the channels.
	go func() {
		wg.Wait()
		close(ch)
		close(chErr)
	}()

	// Collect the results.
	var (
		members = make([]sdkkonnectcomp.Members, 0, len(cp.Spec.Members))
		errs    = make([]error, 0)
	)
	for {
		select {
		case <-ctx.Done():
			break
		case err, ok := <-chErr:
			if ok {
				errs = append(errs, err)
			} else {
				chErr = nil
			}

		case memberID, ok := <-ch:
			if ok {
				members = append(members, sdkkonnectcomp.Members{
					ID: lo.ToPtr(memberID),
				})
			} else {
				ch = nil
			}
		}

		// Break the loop if both channels are closed.
		if ch == nil && chErr == nil {
			break
		}
	}
```

This way we ensure that all values are received from both channels before proceeding.

Reply:
This could be an improvement over what was proposed in the PR but given that we went with using `conc` this is not going to be needed.

Thanks for taking time for looking into this though 🙇